// @protocolink/core/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOGIC_BATCH_TYPED_DATA_TYPES = exports.PERMIT_SIG_DEADLINE = exports.PERMIT_EXPIRATION = exports.OFFSET_NOT_USED = exports.BPS_NOT_USED = void 0;
exports.BPS_NOT_USED = 0;
exports.OFFSET_NOT_USED = '0x8000000000000000000000000000000000000000000000000000000000000000'; // 2^255
exports.PERMIT_EXPIRATION = 2592000; // 30d
exports.PERMIT_SIG_DEADLINE = 86400; // 1d
exports.LOGIC_BATCH_TYPED_DATA_TYPES = {
    LogicBatch: [
        { name: 'logics', type: 'Logic[]' },
        { name: 'fees', type: 'Fee[]' },
        { name: 'referrals', type: 'bytes32[]' },
        { name: 'deadline', type: 'uint256' },
    ],
    Logic: [
        { name: 'to', type: 'address' },
        { name: 'data', type: 'bytes' },
        { name: 'inputs', type: 'Input[]' },
        { name: 'wrapMode', type: 'uint8' },
        { name: 'approveTo', type: 'address' },
        { name: 'callback', type: 'address' },
    ],
    Fee: [
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
        { name: 'metadata', type: 'bytes32' },
    ],
    Input: [
        { name: 'token', type: 'address' },
        { name: 'balanceBps', type: 'uint256' },
        { name: 'amountOrOffset', type: 'uint256' },
    ],
};
